src/psyclone/domain/lfric/lfric_arg_descriptor.py

```diff
@@ -0,0 +1,676 @@
+            raise ParseError(
+                "In the LFRic API each 'meta_arg' entry must have at "
+                "least 2 args, but found '{0}'".format(len(arg_type.args)))
+
+        # Check the first argument descriptor. If it is a binary operator
+        # then it has to be a field vector with an "*n" appended where "*"
+        # is a separator operator and "n > 1" is a vector size. If it is
+        # a variable then it can be other allowed type of argument.
+        argtype = None
+        separator = None
+        if isinstance(arg_type.args[0], expr.BinaryOperator):
+            argtype = arg_type.args[0].toks[0]
+            separator = arg_type.args[0].toks[1]
+        else:
+            argtype = arg_type.args[0]
+
+        # First check for a valid argument type. It has to be a variable
+        # (FunctionVar expression) and have a valid LFRic API argument name.
+        if isinstance(argtype, expr.FunctionVar) and argtype.name in \
+           LFRicArgDescriptor.VALID_ARG_TYPE_NAMES:
+            self._type = argtype.name
+        else:
+            raise ParseError(
+                "In the LFRic API the 1st argument of a 'meta_arg' "
+                "entry should be a valid argument type (one of {0}), but "
+                "found '{1}' in '{2}'".
+                format(LFRicArgDescriptor.VALID_ARG_TYPE_NAMES,
+                       argtype, arg_type))
+
+        # Check for a valid vector size in case of a binary
+        # operator expression
+        if separator:
+            self._validate_vector_size(separator, arg_type)
+
+        # The 2nd arg is an access descriptor
+        # Convert from GH_* names to the generic access type
+        api_config = Config.get().api_conf(API)
+        access_mapping = api_config.get_access_mapping()
+        try:
+            self._access_type = access_mapping[arg_type.args[1].name]
+        except KeyError:
+            valid_names = api_config.get_valid_accesses_api()
+            raise ParseError(
+                "In the LFRic API the 2nd argument of a 'meta_arg' entry "
+                "must be a valid access descriptor (one of {0}), but found "
+                "'{1}' in '{2}'".format(valid_names,
+                                        arg_type.args[1].name, arg_type))
+
+        # Reduction access descriptors are only valid for real scalar
+        # arguments.
+        # TODO in issues #471 and #138: "_validate" methods for fields on
+        # discontinuous function spaces and operators check for the allowed
+        # accesses instead of exceptions so the reduction access check is
+        # redundant for them. However, fields on 'ANY_SPACE's need to
+        # differentiate between looping over cells and DoFs for the allowed
+        # accesses. When this is introduced this check can be moved to
+        # "_validate_scalar" method and repurposed to disallow integer
+        # scalar reductions.
+        if self._type != "gh_real" and self._access_type in \
```

Comment:
This check is now moved to the `_validate_scalar` method where it should have been from the start. My reasoning for not doing it sooner is below.

This is part of the original code and the original checks dealt with accesses on the basis of what is not allowed (individual exceptions). Reduction access for all arguments was caught here rather than checking against the allowed accesses for fields and operators. I indtroduced the "validate" methods in this issue to be able to _define allowed accesses_ for each argument type and check whether the argument access is in these lists. However, I did not do this for the  fields on continuous and `any_space` spaces. They have old-style checks where they just forbid `READWRITE` and rely on this check to deal with the reduction (like all arguments except real scalars in the original implementation). I did not define list of allowed accesses for these fields becuause these fields still use technically incorrect `WRITE` access and defining allowed accesses list would mean including it. I planned to deal with this separately in issues #138 and #471.

However, I think you have a point about being consistent and checking against permitted accesses. Therefore I defined such list for continuous fields which for now includes `WRITE` and referenced #138 and #471.